In [1]:
import pandas as pd
from pathlib import Path

## 1. Load

In [2]:
xlsx_file = pd.read_excel("../data/raw/fuelcheck_pricehistory_dec2025.xlsx")
xlsx_file.head()

,ServiceStationName,Address,Suburb,Postcode,Brand,FuelCode,PriceUpdatedDate,Price
0,BP RANDWICK,"8-10 Perouse Road, RANDWICK NSW 2031",RANDWICK,2031,BP,P98,2025-12-01 01:27:10,199.9
1,Metro Petroleum South Hurstville,"928-930 KING GEORGES RD, BLAKEHURST NSW 2221",BLAKEHURST,2221,Metro Fuel,E10,2025-12-01 02:07:54,164.9
2,Metro Petroleum South Hurstville,"928-930 KING GEORGES RD, BLAKEHURST NSW 2221",BLAKEHURST,2221,Metro Fuel,U91,2025-12-01 02:07:54,166.9
3,Metro Petroleum South Hurstville,"928-930 KING GEORGES RD, BLAKEHURST NSW 2221",BLAKEHURST,2221,Metro Fuel,PDL,2025-12-01 02:07:54,174.9
4,ENHANCE TERALBA,"63 York St, Teralba NSW 2284",Teralba,2284,Enhance,U91,2025-12-01 04:16:26,179.9


In [3]:
csv_file = pd.read_csv("../data/raw/price_history_checks_oct2025.csv")
csv_file.head()

,ServiceStationName,Address,Suburb,Postcode,Brand,FuelCode,PriceUpdatedDate,Price
0,7-Eleven Fairfield,"320 Polding St, FAIRFIELD WEST NSW 2165",FAIRFIELD WEST,2165,7-Eleven,E10,1/10/2025 0:00,163.9
1,7-Eleven Fairfield,"320 Polding St, FAIRFIELD WEST NSW 2165",FAIRFIELD WEST,2165,7-Eleven,U91,1/10/2025 0:00,165.9
2,7-Eleven Fairfield,"320 Polding St, FAIRFIELD WEST NSW 2165",FAIRFIELD WEST,2165,7-Eleven,P95,1/10/2025 0:00,180.9
3,7-Eleven Fairfield,"320 Polding St, FAIRFIELD WEST NSW 2165",FAIRFIELD WEST,2165,7-Eleven,P98,1/10/2025 0:00,188.9
4,7-Eleven Berowra,"965-969 Pacific Highway (Corner Waratah Road),...",Berowra,2081,7-Eleven,P95,1/10/2025 0:00,181.9


In [4]:
folder_path = Path("../data/raw")

merge_file = []

for file_path in sorted(folder_path.iterdir()):
    if file_path.suffix.lower() == ".csv":
        merge_file.append(pd.read_csv(file_path, dtype=str, keep_default_na=False).assign(source_file=file_path.name))
    elif file_path.suffix.lower() == ".xlsx":
        merge_file.append(pd.read_excel(file_path, dtype=str, keep_default_na=False).assign(source_file=file_path.name))

full_table = pd.concat(merge_file, ignore_index=True)
full_table.head()

,ServiceStationName,Address,Suburb,Postcode,Brand,FuelCode,PriceUpdatedDate,Price,source_file
0,Ampol Denham Court,"194 CAMPBELLTOWN RD, DENHAM COURT NSW 2565",DENHAM COURT,2565,Ampol,U91,2026-04-01 00:06:24,239.5,fuelcheck_pricehistory_apr2026.xlsx
1,Liberty Albert,"9, Melrose Road, Albert NSW 2873",Albert,2873,Liberty,U91,2026-04-01 00:22:59,274.9,fuelcheck_pricehistory_apr2026.xlsx
2,Liberty Bourke,"96-104 ANSON ST, BOURKE NSW 2840",BOURKE,2840,Liberty,U91,2026-04-01 00:23:27,264.9,fuelcheck_pricehistory_apr2026.xlsx
3,Liberty Bourke,"96-104 ANSON ST, BOURKE NSW 2840",BOURKE,2840,Liberty,P95,2026-04-01 00:23:27,274.9,fuelcheck_pricehistory_apr2026.xlsx
4,SHELL TAMWORTH TRUCKSTOP (unmanned),"9-13 Gunnedah Road, Tamworth NSW 2340",Tamworth,2340,Shell,DL,2026-04-01 00:23:53,299.9,fuelcheck_pricehistory_apr2026.xlsx


## 2.Profile

### Volume

- 27.8s for loading 13 files
- full_table have 1258332 rows

In [5]:
# Rows for each month
full_table.groupby("source_file").size()

source_file
fuelcheck_pricehistory_apr2026.xlsx     126651
fuelcheck_pricehistory_aug2025.xlsx      74898
fuelcheck_pricehistory_aug2026.xlsx     129143
fuelcheck_pricehistory_dec2025.xlsx      88277
fuelcheck_pricehistory_jan2026.xlsx      74525
fuelcheck_pricehistory_july2026.xlsx    155167
fuelcheck_pricehistory_jun2026.xlsx     103735
fuelcheck_pricehistory_may2026.xlsx      76567
fuelcheck_pricehistory_nov2025.xlsx      79181
fuelcheck_pricehistory_sep2025.xlsx      66938
price_history_checks_feb2026.csv         66981
price_history_checks_mar2026.csv        147137
price_history_checks_oct2025.csv         69132
dtype: int64

July 2026 is the largest and September 2025 is the smallest

In [6]:
full_table.loc[full_table["source_file"] == "fuelcheck_pricehistory_july2026.xlsx"]

,ServiceStationName,Address,Suburb,Postcode,Brand,FuelCode,PriceUpdatedDate,Price,source_file
493494,BP Mascot,"1077 Botany Road, Mascot NSW 2020",Mascot,2020,BP,P98,2026-07-01 00:01:54,168.9,fuelcheck_pricehistory_july2026.xlsx
493495,BP Mascot,"1077 Botany Road, Mascot NSW 2020",Mascot,2020,BP,U91,2026-07-01 00:01:54,169.9,fuelcheck_pricehistory_july2026.xlsx
493496,BP Mascot,"1077 Botany Road, Mascot NSW 2020",Mascot,2020,BP,PDL,2026-07-01 00:01:54,189.9,fuelcheck_pricehistory_july2026.xlsx
493497,BP Mascot,"1077 Botany Road, Mascot NSW 2020",Mascot,2020,BP,P98,2026-07-01 00:02:14,189.9,fuelcheck_pricehistory_july2026.xlsx
493498,Medco Kanahooka,"90-94 Lakeside Dr, KANAHOOKA NSW 2530",KANAHOOKA,2530,Independent,U91,2026-07-01 00:02:32,179.7,fuelcheck_pricehistory_july2026.xlsx
...,...,...,...,...,...,...,...,...,...
648656,7-Eleven Ryde,"326-328 Blaxland Road, Ryde NSW 2112",Ryde,2112,7-Eleven,E10,2026-07-31 23:24:14,191.9,fuelcheck_pricehistory_july2026.xlsx
648657,7-Eleven Ryde,"326-328 Blaxland Road, Ryde NSW 2112",Ryde,2112,7-Eleven,U91,2026-07-31 23:24:14,193.9,fuelcheck_pricehistory_july2026.xlsx
648658,Metro Petroleum Ingleburn,"21-23 CUMBERLAND RD, INGLEBURN NSW 2565",INGLEBURN,2565,Metro Fuel,E10,2026-07-31 23:24:54,189.5,fuelcheck_pricehistory_july2026.xlsx
648659,Bp woonona,"426 Princes Highway, WOONONA NSW 2517",WOONONA,2517,BP,U91,2026-07-31 23:56:07,195.9,fuelcheck_pricehistory_july2026.xlsx


### Completeness

In [7]:
# Make a table to check missing, blank and placeholder values inside each column in the table
blank_space = []
for col in full_table.columns:
    if col == "source_file":
        continue
        
    column = col
    missing = full_table[col].isna().sum()
    blank = full_table[col].str.strip().eq("").sum()
    placeholder = full_table[col].str.strip().str.lower().isin(["-", "n/a", "null", "?", ".", "none", "nan", "na", "unknown"]).sum()
    symbol_only = full_table[col].str.contains(r"^[^a-zA-Z0-9]+$", regex=True).sum()
    changed_by_trim = (full_table[col] != full_table[col].str.strip()).sum()
    dictionary = {"column": column,
                  "missing": missing,
                  "blank": blank,
                  "placeholder":placeholder,
                  "symbol_only": symbol_only,
                  "changed_by_trim": changed_by_trim}
    
    blank_space.append(dictionary)
    
summary_table = pd.DataFrame(blank_space).set_index("column")

In [8]:
summary_table

,missing,blank,placeholder,symbol_only,changed_by_trim
column,,,,,
ServiceStationName,0,0,0,0,0
Address,0,0,0,0,2
Suburb,0,0,0,0,0
Postcode,0,0,0,0,0
Brand,0,0,0,0,0
FuelCode,0,0,0,0,0
PriceUpdatedDate,0,0,0,0,0
Price,0,0,0,0,0


In [9]:
full_table.loc[full_table["Address"] != full_table["Address"].str.strip()]

,ServiceStationName,Address,Suburb,Postcode,Brand,FuelCode,PriceUpdatedDate,Price,source_file
128369,Enhance Benandarah,"10876 Princes Hwy, Benandarah NSW 2536",Benandarah,2536,Enhance,U91,2025-08-02 13:43:50,174.9,fuelcheck_pricehistory_aug2025.xlsx
128370,Enhance Benandarah,"10876 Princes Hwy, Benandarah NSW 2536",Benandarah,2536,Enhance,P98,2025-08-02 13:43:50,193.9,fuelcheck_pricehistory_aug2025.xlsx


2 rows, Aug 2025, address have white space, trim in cleaning

#### Summary

- Method: 5 checks per column (missing, blank after trimming, known placeholder after lowecasing, symbol_only values, and changed_by_trim).
- Result: no missing, blank, placeholder, symbol_only in any of the 8 columns.
- Finding: 2 `Address` rows with whitespace.
- Conclusion: Data is complete. No rows need removing for missing values

### Consistency

#### PriceUpdatedDate

In [10]:
# Replace digits with '#' to expose the format structure
formats = (full_table["PriceUpdatedDate"]
           .astype(str)
           .str.replace(r'[a-zA-Z]', 'w', regex=True)
           .str.replace(r'\d', '#', regex=True))
formats.unique()

<ArrowStringArray>
['####-##-## ##:##:##',      '#/##/#### #:##',     '#/##/#### ##:##',
     '##/##/#### #:##',    '##/##/#### ##:##']
Length: 5, dtype: str

In [11]:
pd.crosstab(full_table["source_file"], formats, margins=True)

PriceUpdatedDate,####-##-## ##:##:##,##/##/#### ##:##,##/##/#### #:##,#/##/#### ##:##,#/##/#### #:##,All
source_file,,,,,,
fuelcheck_pricehistory_apr2026.xlsx,126651,0,0,0,0,126651
fuelcheck_pricehistory_aug2025.xlsx,74898,0,0,0,0,74898
fuelcheck_pricehistory_aug2026.xlsx,129143,0,0,0,0,129143
fuelcheck_pricehistory_dec2025.xlsx,88277,0,0,0,0,88277
fuelcheck_pricehistory_jan2026.xlsx,74525,0,0,0,0,74525
fuelcheck_pricehistory_july2026.xlsx,155167,0,0,0,0,155167
fuelcheck_pricehistory_jun2026.xlsx,103735,0,0,0,0,103735
fuelcheck_pricehistory_may2026.xlsx,76567,0,0,0,0,76567
fuelcheck_pricehistory_nov2025.xlsx,79181,0,0,0,0,79181


In [12]:
# Make 5 different into 2 groups: ISO (containing `-`) and slash (containing `/`)
formats_group = {"####-##-## ##:##:##":"ISO", '#/##/#### #:##': "slash", '#/##/#### ##:##': "slash", '##/##/#### #:##': "slash", '##/##/#### ##:##': "slash"}
formats_table = pd.crosstab(full_table["source_file"], formats.map(formats_group), margins=True)
formats_table

PriceUpdatedDate,ISO,slash,All
source_file,,,
fuelcheck_pricehistory_apr2026.xlsx,126651,0,126651
fuelcheck_pricehistory_aug2025.xlsx,74898,0,74898
fuelcheck_pricehistory_aug2026.xlsx,129143,0,129143
fuelcheck_pricehistory_dec2025.xlsx,88277,0,88277
fuelcheck_pricehistory_jan2026.xlsx,74525,0,74525
fuelcheck_pricehistory_july2026.xlsx,155167,0,155167
fuelcheck_pricehistory_jun2026.xlsx,103735,0,103735
fuelcheck_pricehistory_may2026.xlsx,76567,0,76567
fuelcheck_pricehistory_nov2025.xlsx,79181,0,79181


In [13]:
# second number of "PriceUpdatedDate" in October 2025 is month
oct2025 = (full_table
                .loc[(full_table["source_file"] == "price_history_checks_oct2025.csv"),["source_file", "PriceUpdatedDate"]]
                .assign(month = lambda x: x["PriceUpdatedDate"].str.split("/").str[1])
          )
oct2025["month"].unique()

array(['10'], dtype=object)

In [14]:
# first number go above 12 so the first number is day
oct2025 = (full_table
                .loc[(full_table["source_file"] == "price_history_checks_oct2025.csv"),["source_file", "PriceUpdatedDate"]]
                .assign(day = lambda x: x["PriceUpdatedDate"].str.split("/").str[0])
          )
oct2025["day"].unique()

array(['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12',
       '13', '14', '15', '16', '17', '18', '19', '20', '21', '22', '23',
       '24', '25', '26', '27', '28', '29', '30', '31'], dtype=object)

Summary:
- 2 Formats: ISO (1189200 lines in 12 files) and slash (69132 lines in only 1 file: `price_history_checks_oct2025.csv`)
- Not a CSV and Excel files difference: other CSV files except `price_history_checks_oct2025.csv` use ISO format
- Slash format is day first: day/month/year
- Slash format have no seconds
- The risk: Automatic parsing reads slash format as month first, which can cause it gives the wrong dates.
- The proposed fix: parse with a stated format for each source

#### Postcode

In [28]:
# Extract state from address
non_NSW_stations_check1 = (full_table["Address"]
                           .str.upper()
                           .str.findall(r"\b(NSW|ACT|VIC|QLD|SA|WA|TAS|NT|"
                                        r"NEW\s+SOUTH\s+WALES|AUSTRALIAN\s+CAPITAL\s+TERRITORY|VICTORIA|QUEENSLAND|SOUTH\s+AUSTRALIA|WESTERN\s+AUSTRALIA|TASMANIA|NORTHERN\s+TERRITORY)\b"
                                        ).str[-1].replace({"NEW SOUTH WALES": "NSW", 
                                                           "AUSTRALIAN CAPITAL TERRITORY": "ACT",
                                                           "VICTORIA":"VIC",
                                                           "QUEENSLAND": "QLD",
                                                           "SOUTH AUSTRALIA": "SA",
                                                           "WESTERN AUSTRALIA": "WA",
                                                           "TASMANIA": "TAS",
                                                           "NORTHERN TERRITORY": "NT"})
                           )
non_NSW_stations_check1.value_counts()

Address
NSW    1227369
ACT      30963
Name: count, dtype: int64

In [ ]:
# Confirm if all Station are in NSW
non_NSW_stations_check2 = (
                            (full_table["Postcode"].astype("int16").between(1000, 1999)) | 
                            (full_table["Postcode"].astype("int16").between(2000, 2599)) | 
                            (full_table["Postcode"].astype("int16").between(2619, 2899)) | 
                            (full_table["Postcode"].astype("int16").between(2921, 2999))
                          ).map({True:"NSW", False:"Not NSW"})

non_NSW_stations_check2

0          NSW
1          NSW
2          NSW
3          NSW
4          NSW
          ... 
1258327    NSW
1258328    NSW
1258329    NSW
1258330    NSW
1258331    NSW
Name: Postcode, Length: 1258332, dtype: str

In [29]:
pd.crosstab(index=non_NSW_stations_check1, columns=non_NSW_stations_check2)

Postcode,NSW,Not NSW
Address,,
ACT,590,30373
NSW,1226967,402


In [27]:
non_NSW_stations_check1.unique()

array(['NSW', 'ACT'], dtype=object)

In [45]:
act_address_nsw_postcode = ((non_NSW_stations_check1 == "ACT") & (non_NSW_stations_check2 == "NSW"))
full_table.loc[act_address_nsw_postcode, ["ServiceStationName", "Address", "Suburb", "Postcode"]]

,ServiceStationName,Address,Suburb,Postcode
2922,Ampol Hume,"96 Sawmill Cct, Hume ACT 2620",Hume,2620
2923,Ampol Hume,"96 Sawmill Cct, Hume ACT 2620",Hume,2620
5130,EG Ampol Hume,"94-96 Sheppard St, Hume ACT 2620",Hume,2620
5131,EG Ampol Hume,"94-96 Sheppard St, Hume ACT 2620",Hume,2620
5132,EG Ampol Hume,"94-96 Sheppard St, Hume ACT 2620",Hume,2620
...,...,...,...,...
1254095,Ampol Hume,"96 Sawmill Cct, Hume ACT 2620",Hume,2620
1254301,EG Ampol Hume,"94-96 Sheppard St, Hume ACT 2620",Hume,2620
1254302,EG Ampol Hume,"94-96 Sheppard St, Hume ACT 2620",Hume,2620
1254303,EG Ampol Hume,"94-96 Sheppard St, Hume ACT 2620",Hume,2620


In [44]:
nsw_address_non_nsw_postcode = ((non_NSW_stations_check1 == "NSW") & (non_NSW_stations_check2 == "Not NSW"))
full_table.loc[nsw_address_non_nsw_postcode, ["ServiceStationName", "Address", "Suburb", "Postcode"]]

,ServiceStationName,Address,Suburb,Postcode
6976,Mobil Jennings,"1823 NEW ENGLAND HWY, JENNINGS NSW 4383",JENNINGS,4383
6977,Mobil Jennings,"1823 NEW ENGLAND HWY, JENNINGS NSW 4383",JENNINGS,4383
6978,Mobil Jennings,"1823 NEW ENGLAND HWY, JENNINGS NSW 4383",JENNINGS,4383
7856,Mobil Jennings,"1823 NEW ENGLAND HWY, JENNINGS NSW 4383",JENNINGS,4383
7857,Mobil Jennings,"1823 NEW ENGLAND HWY, JENNINGS NSW 4383",JENNINGS,4383
...,...,...,...,...
1235218,United Jennings,"Border Roadhouse 1823 New England Hwy, JENNING...",JENNINGS,4383
1255201,United Jennings,"Border Roadhouse 1823 New England Hwy, JENNING...",JENNINGS,4383
1255202,United Jennings,"Border Roadhouse 1823 New England Hwy, JENNING...",JENNINGS,4383
1255203,United Jennings,"Border Roadhouse 1823 New England Hwy, JENNING...",JENNINGS,4383


In [53]:
(full_table
 .loc[act_address_nsw_postcode, ["ServiceStationName", "Address", "Suburb", "Postcode"]]
 .pivot_table(index="ServiceStationName",
              values=["Address", "Suburb", "Postcode"],
              aggfunc="count",
              margins=True)
)

,Address,Postcode,Suburb
ServiceStationName,,,
Ampol Hume,175,175,175
EG Ampol Hume,415,415,415
All,590,590,590


In [52]:
(full_table
 .loc[nsw_address_non_nsw_postcode, ["ServiceStationName", "Address", "Suburb", "Postcode"]]
 .pivot_table(index="ServiceStationName",
              values=["Address", "Suburb", "Postcode"],
              aggfunc="count",
              margins=True)
)

,Address,Postcode,Suburb
ServiceStationName,,,
Barooga General Store,96,96,96
Mobil Jennings,253,253,253
United Jennings,53,53,53
All,402,402,402


#### Brand

In [18]:
full_table["Brand"].value_counts()

Brand
7-Eleven             254037
BP                   149168
Metro Fuel           126962
Ampol Foodary        126946
EG Ampol             116462
                      ...  
The Major                34
24Xpress logo            24
Bribbaree Servo          22
Linmac Co Pty Ltd        12
Eric Brown Motors         8
Name: count, Length: 66, dtype: int64

In [19]:
full_table["Brand"].str.lower().value_counts()

Brand
7-eleven             254037
bp                   149168
metro fuel           126962
ampol foodary        126946
eg ampol             116462
                      ...  
the major                34
24xpress logo            24
bribbaree servo          22
linmac co pty ltd        12
eric brown motors         8
Name: count, Length: 66, dtype: int64